From 6f2ec6f2baa5979209259f8267f2c2b60a79e22c Mon Sep 17 00:00:00 2001
From: Andrew Pope <andrew.pope@morsemicro.com>
Date: Fri, 20 Jun 2025 17:53:09 +1000
Subject: [PATCH] wifi: nl80211: Introduce the channel flag
 IEEE80211_CHAN_IGNORE for BCF disabled channels

In this changeset a a new flag IEEE80211_CHAN_IGNORE, which gets applied to BCF-disabled
channels on boot, is introduced. When userspace retrieves wiphy information via nl80211
this flag is translated to NL80211_FREQUENCY_ATTR_DISABLED. This allows us to side-step
the logic in cfg80211 that considers wider channels unusable (when IEEE80211_CHAN_DISABLED
is applied), while still presenting a channel as being disabled to userspace.

This change will also enforce (in the supplicant & hostap), that a primary (or secondary in
 5G land) channel is not disabled before tuning to it.

Approved-by: Adam Baumann
Approved-by: Dan Callaghan
Approved-by: Pradeep Reddy
Approved-by: Ria Polly Thomas
Approved-by: James Herbert
Approved-by: Joshua Wilkinson
---
 include/net/cfg80211.h | 4 ++++
 net/wireless/nl80211.c | 3 +++
 2 files changed, 7 insertions(+)

diff --git a/include/net/cfg80211.h b/include/net/cfg80211.h
index 5b3a63c377d6..41e750fe5aa3 100644
--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -115,6 +115,9 @@ struct wiphy;
  *	This may be due to the driver or due to regulatory bandwidth
  *	restrictions.
  * @IEEE80211_CHAN_NO_EHT: EHT operation is not permitted on this channel.
+ * @IEEE80211_CHAN_IGNORE: Mark this channel as disabled when
+ *	communicating channel list via nl80211.
+ *
  */
 enum ieee80211_channel_flags {
 	IEEE80211_CHAN_DISABLED		= 1<<0,
@@ -138,6 +141,7 @@ enum ieee80211_channel_flags {
 	IEEE80211_CHAN_16MHZ		= 1<<18,
 	IEEE80211_CHAN_NO_320MHZ	= 1<<19,
 	IEEE80211_CHAN_NO_EHT		= 1<<20,
+	IEEE80211_CHAN_IGNORE		= 1<<21,
 };
 
 #define IEEE80211_CHAN_NO_HT40 \
diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index 545f7ecaad76..197160b1a286 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -1124,6 +1124,9 @@ static int nl80211_msg_put_channel(struct sk_buff *msg, struct wiphy *wiphy,
 	if ((chan->flags & IEEE80211_CHAN_DISABLED) &&
 	    nla_put_flag(msg, NL80211_FREQUENCY_ATTR_DISABLED))
 		goto nla_put_failure;
+	if ((chan->flags & IEEE80211_CHAN_IGNORE) &&
+	    nla_put_flag(msg, NL80211_FREQUENCY_ATTR_DISABLED))
+		goto nla_put_failure;
 	if (chan->flags & IEEE80211_CHAN_NO_IR) {
 		if (nla_put_flag(msg, NL80211_FREQUENCY_ATTR_NO_IR))
 			goto nla_put_failure;
